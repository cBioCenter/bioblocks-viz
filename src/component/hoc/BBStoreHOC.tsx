import * as React from 'react';
import { connect, MapDispatchToProps, MapStateToProps, Provider } from 'react-redux';
import { BBStore } from '~bioblocks-viz~/reducer';

export const connectWithBBStore = <TStateProps, TDispatchProps, TOwnProps, State>(
  mapStateToProps: MapStateToProps<TStateProps, TOwnProps, State>,
  mapDispatchToProps: MapDispatchToProps<TDispatchProps, TOwnProps>,
  WrappedComponent: React.ComponentType<any>,
) => {
  const connectedComponent = connect(mapStateToProps, mapDispatchToProps)(WrappedComponent);

  return withBBStore(connectedComponent);
};

export const withBBStore = <P extends object>(WrappedComponent: React.ComponentType<P>) => (props: P) => {
  WrappedComponent.displayName = `withBBStore(${WrappedComponent.displayName})`;

  return (
    <Provider store={BBStore}>
      <WrappedComponent {...props} />
    </Provider>
  );
};
