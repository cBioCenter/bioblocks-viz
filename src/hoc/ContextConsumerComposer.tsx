import * as React from 'react';

/**
 * ContextConsumerComposer is meant to streamline connecting various context consumers.
 * Heavily based off of [react-composer](https://github.com/jamesplease/react-composer).
 *
 * However, this only accepts an array of context types instead.
 */
export interface IComposerProps {
  components: Array<React.ComponentType<any>>;
  children(args: any[]): JSX.Element;
}

class ContextConsumerComposer extends React.Component<IComposerProps, any> {
  public render() {
    const { children, components } = this.props;

    return renderRecursive(children, components);
  }
}

/**
 * Recursively build up elements from props.components and accumulate `results` along the way.
 */
const renderRecursive = (
  render: (args: any) => React.ReactElement<any>,
  remaining: Array<React.ComponentType<any>>,
  results: Array<React.ReactElement<any>> = [],
): React.ReactElement<any> => {
  // Once components is exhausted, we can render out the results array.
  if (!remaining[0]) {
    return render(results);
  }

  // Continue recursion for remaining items.
  // results.concat([value]) ensures [...results, value] instead of [...results, ...value]
  function nextRender(value: any) {
    return renderRecursive(render, remaining.slice(1), results.concat([value]));
  }

  const Item = remaining[0];

  // When it is an element, enhance the element's props with the render prop.
  return React.cloneElement(<Item>{nextRender}</Item>, { displayName: Item.displayName });
};

export { ContextConsumerComposer };
