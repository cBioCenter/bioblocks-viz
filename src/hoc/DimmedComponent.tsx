import * as React from 'react';
import { Dimmer, Loader } from 'semantic-ui-react';

export const withDimmedLoader = (component: JSX.Element, active: boolean) => (
  <Dimmer.Dimmable dimmed={true}>
    <Dimmer active={active}>
      <Loader />
    </Dimmer>
    {component}
  </Dimmer.Dimmable>
);
