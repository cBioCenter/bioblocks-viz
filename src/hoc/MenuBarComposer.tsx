import * as React from 'react';

import { Card, Menu } from 'semantic-ui-react';

// Options for the HOC factory that are not dependent on props values
interface IMenuBarProps {
  showMenu?: boolean;
}

const withMenuBar = ({ showMenu = true }: IMenuBarProps = {}) => <OriginalProps extends {}>(
  WrappedComponent: React.ComponentType<OriginalProps & {}>,
): React.ComponentClass<OriginalProps> => {
  return class HOC extends React.Component<OriginalProps & {}> {
    public render() {
      return (
        <div>
          {showMenu && <Menu>{WrappedComponent.displayName}</Menu>}
          <Card>
            <WrappedComponent {...this.props} />
          </Card>
        </div>
      );
    }
  };
};

// tslint:disable-next-line:export-name
export { withMenuBar };
