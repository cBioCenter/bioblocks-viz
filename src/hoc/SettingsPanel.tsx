import * as React from 'react';
import { Button, Grid, Menu, Sidebar } from 'semantic-ui-react';

import { IContactMapConfiguration } from '../component/ContactMap';
import { ChellRadioGroup } from '../component/widget/ChellRadioGroup';
import { ChellSlider } from '../component/widget/ChellSlider';
import { CONFIGURATION_COMPONENT_TYPE, CONTACT_DISTANCE_PROXIMITY } from '../data/chell-data';

export interface ISettingsPanelState {
  visible: boolean;
}

export interface ISettingsPanelProps {
  configurations: IContactMapConfiguration[];
  width: number;
}

const withSettingsPanel = <P extends object>(WrappedComponent: React.ComponentType<P>) =>
  class SettingsPanel extends React.Component<P & ISettingsPanelProps, ISettingsPanelState> {
    public static defaultProps = {
      configurations: new Array<IContactMapConfiguration>({
        name: 'Measuring Proximity',
        onChange: () => {
          return;
        },
        type: CONFIGURATION_COMPONENT_TYPE.RADIO,
        values: {
          current: Object.values(CONTACT_DISTANCE_PROXIMITY).indexOf(CONTACT_DISTANCE_PROXIMITY.CLOSEST),
          max: 0,
          min: 1,
          options: Object.values(CONTACT_DISTANCE_PROXIMITY),
        },
      }),
    };

    constructor(props: any) {
      super(props);
      this.state = {
        visible: false,
      };
    }

    public render() {
      const { visible } = this.state;
      const style: React.CSSProperties = { width: 400 * 0.9 };

      return (
        <div>
          <Grid columns={1}>
            <Grid.Column>{this.renderSettingsButton()}</Grid.Column>
            <Sidebar.Pushable>
              <Sidebar
                as={Menu}
                animation="overlay"
                inverted={false}
                style={{ opacity: 0.5, width: '100%' }}
                vertical={true}
                visible={visible}
              >
                {this.renderConfigurations(this.props.configurations, style)}
              </Sidebar>

              <Sidebar.Pusher>
                <WrappedComponent {...this.props} />
              </Sidebar.Pusher>
            </Sidebar.Pushable>
          </Grid>
        </div>
      );
    }

    public onButtonClick = (e: React.MouseEvent) =>
      this.setState({
        visible: !this.state.visible,
      });

    public renderSettingsButton = () => (
      <Button icon={'settings'} basic={true} floated={'right'} onClick={this.onButtonClick} />
    );

    public renderConfigurations(
      configurations: IContactMapConfiguration[],
      style: React.CSSProperties[] | React.CSSProperties,
    ) {
      return configurations.map(config => {
        const id = config.name
          .toLowerCase()
          .split(' ')
          .join('-');
        switch (config.type) {
          case CONFIGURATION_COMPONENT_TYPE.SLIDER:
            return this.renderConfigurationSlider(config, id, style);
          case CONFIGURATION_COMPONENT_TYPE.RADIO:
            return this.renderConfigurationRadioButton(config, id, style);
        }
      });
    }

    public renderConfigurationSlider(
      config: IContactMapConfiguration,
      id: string,
      sliderStyle: React.CSSProperties[] | React.CSSProperties,
    ) {
      return (
        <ChellSlider
          className={id}
          key={id}
          value={config.values.current}
          label={config.name}
          max={config.values.max}
          min={config.values.min}
          onChange={config.onChange}
          style={sliderStyle}
        />
      );
    }

    public renderConfigurationRadioButton(
      config: IContactMapConfiguration,
      id: string,
      style: React.CSSProperties[] | React.CSSProperties,
    ) {
      return (
        <ChellRadioGroup
          key={`radio-group-${id}`}
          id={id}
          options={config.values.options!}
          onChange={config.onChange}
          style={style}
        />
      );
    }
  };

export { withSettingsPanel };
export default withSettingsPanel;
