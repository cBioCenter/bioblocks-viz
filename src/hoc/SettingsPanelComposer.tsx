import * as React from 'react';

import { Dimmer, Loader } from 'semantic-ui-react';
import { SettingsPanel } from '~chell-viz~/component';
import { ChellWidgetConfig } from '~chell-viz~/data';

interface IWrappedSettingsProps {
  configurations: ChellWidgetConfig[];
  isDataLoading: boolean;
  showConfigurations: boolean;
  width: number | string;
}
// @ts-ignore
const withSettingsPanel = <P extends object, S extends object>(
  WrappedComponent: React.ComponentType<P>,
  deriveConfigurations?: (component: any) => ChellWidgetConfig[],
) =>
  class extends React.Component<IWrappedSettingsProps> {
    public static WrappedComponent = WrappedComponent;

    public render() {
      const { configurations, isDataLoading, showConfigurations, ...remainingProps } = this.props;

      return (
        <Dimmer.Dimmable dimmed={true}>
          <Dimmer active={isDataLoading}>
            <Loader />
          </Dimmer>
          <SettingsPanel
            configurations={deriveConfigurations !== undefined ? deriveConfigurations(this) : configurations}
            showConfigurations={showConfigurations}
            width={remainingProps.width}
          >
            <WrappedComponent {...remainingProps} />
          </SettingsPanel>
        </Dimmer.Dimmable>
      );
    }
  };

// tslint:disable-next-line:export-name
export { withSettingsPanel };
