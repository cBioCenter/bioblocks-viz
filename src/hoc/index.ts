export * from './ContextConsumerComposer';
export * from './DimmedLoader';
