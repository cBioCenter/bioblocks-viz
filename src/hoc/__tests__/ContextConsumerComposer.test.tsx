import { mount } from 'enzyme';
import * as React from 'react';

import { ContextConsumerComposer } from '~chell-viz~/hoc';

describe('ContextConsumerComposer', () => {
  it('Should allow contexts.', () => {
    const wrapper = mount(<ContextConsumerComposer components={[]}>{() => <div />}</ContextConsumerComposer>);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should allow consumers added directly.', () => {
    const mockContext = React.createContext('foo');
    const wrapper = mount(
      <ContextConsumerComposer components={[mockContext.Consumer]}>{() => <div />}</ContextConsumerComposer>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
