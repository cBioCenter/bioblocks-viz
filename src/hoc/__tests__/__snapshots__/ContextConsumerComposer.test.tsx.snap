// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextConsumerComposer Should allow consumers added directly. 1`] = `
<Component
  components={
    Array [
      Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": "foo",
        "_currentValue2": "foo",
        "unstable_read": [Function],
      },
    ]
  }
>
  <div />
</Component>
`;

exports[`ContextConsumerComposer Should allow consumers added directly. 2`] = `
<Component
  components={
    Array [
      Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": "foo",
        "_currentValue2": "foo",
        "unstable_read": [Function],
      },
    ]
  }
>
  <div />
</Component>
`;

exports[`ContextConsumerComposer Should allow consumers added via function calls. 1`] = `
<Component
  components={
    Array [
      [Function],
    ]
  }
>
  <div />
</Component>
`;

exports[`ContextConsumerComposer Should allow contexts. 1`] = `
<Component
  components={Array []}
>
  <div />
</Component>
`;

exports[`ContextConsumerComposer Should allow contexts. 2`] = `
<Component
  components={Array []}
>
  <div />
</Component>
`;
